projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+	err := json.Unmarshal(file, &dataSlice)
+
+	// if we fail to "unmarshal" the json file then error
+	if err != nil {
+		return errors.New("[3] could not parse any data in JSON format")
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in JSON format")
+	}
+
+	return nil
+}
+
+// [2] Repeated JSON
+func parseRepeatedJSON(file []byte, dataSlice *[]Player) error {
+	// convert the file to a string and then split it on a newline character
+	lines := strings.Split(string(file), "\n")
+
+	// if no lines were parsed then error
+	if len(lines) < 1 {
+		return errors.New("[3] could not parse any data in repeated JSON format")
+	}
+
+	// initialise a player struct to iteratively store each valid player
+	var player Player
+
+	// loop over the slice of strings
+	for _, line := range lines {
+
+		// if the line is empty keep going
+		if len(line) == 0 {
+			continue
+		}
+
+		// if the line starts with a # it is a comment so ignore it ((!)use single quotes for characters)
+		if line[0] == '#' {
+			continue
+		}
+
+		// try to parse the player struct from the line and store it in the player struct
+		err := json.Unmarshal([]byte(line), &player)
+
+		// if we can't parse a player struct from the line then error but keep going
+		if err != nil {
+			continue
+		}
+
+		// add the player struct to the slice of players
+		*dataSlice = append(*dataSlice, player)
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in repeated JSON format")
+	}
+
+	return nil
+}
+
+// [3] CSV
+func parseCSV(file []byte, dataSlice *[]Player) error {
+	// because of the way we have tried to use file []byte as a parameter..
+	// it means we have to use bytes.NewReader on the file... this is JANKY
+	reader := csv.NewReader(bytes.NewReader(file))
+
+	// try to read the lines from the csv file
+	lines, err := reader.ReadAll()
+
+	// if we can't read the lines from the csv file then error
+	if err != nil {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	if len(lines) == 0 {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	// loop over each line
+	for _, line := range lines {
+
+		// (!!!) if we try to read one of the binary files it gives us this: [[��] [AyaPrisha����CharlieMargot]]
+		// so we need to add an additional check here:
+
+		// if the line does not have 2 values (highscore, name) then skip
+		// ((!) this is brittle if a line has valid data but random extra values etc)
+		if len(line) != 2 {
+			continue
+		}
+
+		// get the player name
+		name := line[0]
+
+		// the player has no name so skip
+		if len(name) < 1 {
+			continue
+		}
+
+		//  parse the string to get the score ((!) this is of type int not int32)
+		highScore, err := strconv.Atoi(line[1])
+
+		// if the player has no high score then skip
+		if err != nil {
```

Comment:
Ignoring errors that you come across when parsing data generally leads to bugs :( Particularly silently ignoring them by not even logging that you're doing so.

Imagine you had a bug in your code where you calling a function that assumed all the numbers were positive, but half of the scores in the data file were negative. If you ran your code against it, you would silently get reasonable-seeming results, just ignoring half the rows. This is often _much worse_ than your program returning an error and you needing to investigate it, or your program logging about a warning for you to investigate.

Reply:
I am not sure what to do here
Is it better to "parse what you can" or just immediately error when a single line is not valid...
I am on the fence about that^

Reply:
I added a Stderr log
And funnily enough that gave me a warning error when parsing line 1 of the csv
Which I didn't even think about, since it is the column names
Maybe this will help me to deal with the other point you made about my indexing assumptions